### PR TITLE
[JShellAPI][Testing] Setting up First Integration Test - Simple code snippet evaluation scenario

### PR DESCRIPTION
**Pull request by firasrg**
[contributing]: https://github.com/Together-Java/TJ-Bot/wiki/Contributing
[code_guidelines]: https://github.com/Together-Java/TJ-Bot/wiki/Code-Guidelines
[new_issue]: https://github.com/Together-Java/TJ-Bot/issues/new/choose

## Pull-request

- [ ] I have read the [contributing guidelines][contributing].
- [x] I have read the [code guidelines][code_guidelines].
- [ ] I have created a relating [issue][new_issue].

### Changes

- [x] Some changes in gradle files to 
- [x] New integration test

## Description
This PR is about setting up a first working integration test for JShellAPI only, it contains the following :
- New global property in root project'd gradle.build file to share JShellWrapper's image name across subprojects gradle.build files and Java files as well.
- [JShellAPI] New custom tasks in gradle.build file for docker image : build + remove;
- [JShellAPI] New Integration tests class with a 1st working test on WS `POST/jshel/eval` .

_Note: this PR is about setting the roadmap for testing strategy for JShellAPI only. It doesn't close [the relevant issue](https://github.com/Together-Java/JShellPlaygroundBackend/issues/30)._

### CODE REVIEW

**Comment by Alathreon** [comment]
I don't have time to do a review, so I'll just write here.
There are two problems :
- Your test is too complicated to test something too simple, you need to simplify it, create methods or classes for this so your test can be as simple as for example `assertEquals(sendEval("2+2"), new JShellResult(...))`
- Also include a test to test state. For example, you can send an eval `"var list = List.of(1, 2, 3)"` and then send a separate eval like `"println(list)"`

**Comment by firasrg** [comment]
Hello @Alathreon !

> Your test is too complicated to test something too simple
 I think that my test code is pretty simple, it uses WebTestClient API to establish test in a smooth way (ig it's a good practice as mentioned in [spring docs](https://docs.spring.io/spring-framework/reference/testing/webtestclient.html#webtestclient-tests)) : 
 
 
> create methods or classes for this so your test can be as simple as for example assertEquals(sendEval("2+2"), new JShellResult(...))

I'm not certain i understand this sorry, can u elaborate ?

File: JShellAPI/README.MD [hunk 1]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
 ### jshellapi.dockerCPUsUsage
 The cpu configuration of each container, see [--cpus option of docker](https://docs.docker.com/config/containers/resource_constraints/#cpu).
 ### jshellapi.schedulerSessionKillScanRate
-The rate at which the session killer will check and delete session, in seconds, see [Session timeout](#Session-timeout).
\ No newline at end of file
+The rate at which the session killer will check and delete session, in seconds, see [Session timeout](#Session-timeout).
+
+## Testing
+
+>  The work on testing was made in collaboration with [Alathreon](https://github.com/Alathreon) and [Wazei](https://github.com/tj-wazei). I'd like thank both of them for their trust. - FirasRG
```
Comment (Alathreon) [inline JShellAPI/README.MD:108]: There is no need to say that, readme is supposed to rapidly indicate how to use this project.

File: JShellAPI/README.MD [hunk 2]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
...
+
+This section outlines the work done to set up the first integration test that evaluates Java code by running it in a [Docker](https://www.docker.com/get-started/) container. The test ensures that the [Eval endpoint](#eval) can execute code within the containerized environment of [**JShellWrapper**](../JShellWrapper).
```
Comment (Alathreon) [inline JShellAPI/README.MD:110]: > This section outlines the work done

Don't outline a storyline, just say facts.
And this sentences don't say much, it's obvious what tests do.

File: JShellAPI/README.MD [hunk 3]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
...
+
+### Usage
+
+```java
+@ContextConfiguration(classes = Main.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class JShellApiTests {
```
Comment (Alathreon) [inline JShellAPI/README.MD:117]: You are supposed to explain how to use this project, not write a wiki about how do contribute each feature.

File: JShellAPI/README.MD [hunk 4]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
...
+        
+        // performing a second code execution test ...
+    }
+    // some methods ...
+}
+```
+
+### 1. Java Test Setup
+
+The [@SpringBootTest](https://docs.spring.io/spring-boot/api/java/org/springframework/boot/test/context/SpringBootTest.html) and [@ContextConfiguration](https://docs.spring.io/spring-framework/reference/testing/annotations/integration-spring/annotation-contextconfiguration.html) annotations are needed to prepare the app to tests, like in a real scenario.
```
Comment (Alathreon) [inline JShellAPI/README.MD:144]: What's the point of saying that ?

File: JShellAPI/README.MD [hunk 5]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
...
+
+NOTE: _Test classes must be located under `/src/test/java/{org.togetherjava.jshellapi}`._
```
Comment (Alathreon) [inline JShellAPI/README.MD:146]: That's obvious, otherwise gradle won't work.

File: JShellAPI/README.MD [hunk 6]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
...
+
+- The test uses [WebTestClient](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/test/web/reactive/server/WebTestClient.html) to make HTTP calls to the target endpoint.
```
Comment (Alathreon) [inline JShellAPI/README.MD:148]: You are just quoting your own code, all of this is pointless.

File: JShellAPI/README.MD [hunk 7]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
...
+- Multiple API calls are made within the test method, so a utility instance method was created for reuse.
+- The test ensures that code is correctly evaluated inside the **JShellWrapper** container.
+
+### 2. Gradle Configuration for Tests
+
+The `build.gradle` of this project has been updated to handle **JShellWrapper** Docker image lifecycle during tests.
+
+- **JShellWrapper Image Name**: the image name is injected from the root [build.gradle](../build.gradle) file, to this project's [build.gradle](build.gradle) file and also to [application.yaml](src/main/resources/application.yaml)!
```
Comment (Alathreon) [inline JShellAPI/README.MD:156]: This is the wrong place to write doc, and if you need to specify config, you can specify above, in the corresponding section.

File: JShellAPI/README.MD [hunk 8]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
...
+    group = 'docker'
+    description = 'removes jshellwrapper image'
+    commandLine 'docker', 'rmi', '-f', jshellWrapperImageName
+}
+
+test {
+    dependsOn taskBuildDockerImage
+    finalizedBy taskRemoveDockerImage
+}
+```
```
Comment (Alathreon) [inline JShellAPI/README.MD:185]: There is no need to copy part of the code.

File: JShellAPI/README.MD [hunk 9]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
...
+
+Below are the key dependencies that were added or modified in the `build.gradle` file of this project :
+
+```groovy
+testImplementation('org.springframework.boot:spring-boot-starter-test') {
+  exclude group: 'ch.qos.logback', module: 'logback-classic'
+}
+testImplementation 'org.springframework.boot:spring-boot-starter-webflux'
+```
```
Comment (Alathreon) [inline JShellAPI/README.MD:194]: There is no need to specify this.

File: JShellAPI/README.MD [hunk 10]
```diff
@@ -101,4 +101,97 @@ The maximum ram allocated per container, in megabytes.
...
+
+- The `logback-classic` has been excluded because of an issue encountered when running tests. The issue is typically about a conflict between some dependencies (This solution has been brought based on [a _good_ answer on Stackoverflow](https://stackoverflow.com/a/42641450/10000150))
+- The `spring-boot-starter-webflux` was needed in order to be able to use **WebTestClient**.
```
Comment (Alathreon) [inline JShellAPI/README.MD:197]: That's should be a comment, not in the readme.

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/rest/ApiEndpoints.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package org.togetherjava.jshellapi.rest;
+
+/**
+ * This class holds endpoints mentioned in controllers. The main objective is to keep endpoints
+ * synchronized with testing classes.
+ *
+ * @author Firas Regaieg
```
Comment (Alathreon) [inline JShellAPI/src/main/java/org/togetherjava/jshellapi/rest/ApiEndpoints.java:7]: No need to put authors, we have git history.

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java [hunk 12]
```diff
@@ -29,6 +31,9 @@ public class DockerService implements DisposableBean {
 
     private final DockerClient client;
 
+    @Value("${jshell-wrapper.image-name}")
```
Comment (Alathreon) [inline JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java:34]: Use Config object in the constructor.

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java [hunk 13]
```diff
@@ -59,22 +64,33 @@ private void cleanupLeftovers(UUID currentId) {
 
     public String spawnContainer(long maxMemoryMegs, long cpus, @Nullable String cpuSetCpus,
             String name, Duration evalTimeout, long sysoutLimit) throws InterruptedException {
-        String imageName = "togetherjava.org:5001/togetherjava/jshellwrapper";
+        String imageName = Optional.ofNullable(this.jshellWrapperImageName)
```
Comment (Alathreon) [inline JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java:67]: Use Objects.requireNonNull instead of abusing Optional.

File: JShellAPI/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 14]
```diff
@@ -0,0 +1,8 @@
+{
+  "properties": [
+    {
+      "name": "jshell-wrapper.image-name",
+      "type": "java.lang.String",
+      "description": "JShellWrapper image name injected from the top-level gradle build file."
+  }
+] }
```
Comment (Alathreon) [inline JShellAPI/src/main/resources/META-INF/additional-spring-configuration-metadata.json:8]: What is this ?

File: JShellAPI/src/main/resources/application.yaml [hunk 15]
```diff
@@ -20,6 +20,9 @@ jshellapi:
   dockerResponseTimeout: 60
   dockerConnectionTimeout: 60
 
+jshell-wrapper:
+  image-name: ${jShellWrapperImageName}
+
```
Comment (Alathreon) [inline JShellAPI/src/main/resources/application.yaml:25]: Put it in jshellapi

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+import java.time.Duration;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * This class holds integration tests for JShellAPI. It depends on gradle building image task, fore
+ * more information check "test" section in gradle.build file.
+ *
+ * @author Firas Regaieg
```
Comment (Alathreon) [inline JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:25]: No need to put author, we have git history.

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+import org.togetherjava.jshellapi.rest.ApiEndpoints;
+
+import java.time.Duration;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * This class holds integration tests for JShellAPI. It depends on gradle building image task, fore
+ * more information check "test" section in gradle.build file.
```
Comment (Alathreon) [inline JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:23]: > This class holds integration tests for JShellAPI.

use "Integrates tests for JShellAPI." instead. Don't use the term class, it's redondant.
> fore

for*
>   check "test" section in gradle.build file.

Why do you need to specify this ?

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        final String secondCodeExpression = "a * 2";
+
+        final JShellSnippetResult secondCodeSnippet = new JShellSnippetResult(SnippetStatus.VALID,
+                SnippetType.ADDITION, 2, secondCodeExpression, "8");
+
+        final JShellResult secondCodeExpectedResult =
+                getJShellResultDefaultInstance(secondCodeSnippet);
+
+        assertThat(testEval(testEvalId, secondCodeExpression)).isEqualTo(secondCodeExpectedResult);
+    }
```
Comment (Alathreon) [inline JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:62]: Why do you skip every even line, you are wasting space for no reason.
This whole code could be done in 2 or 4 lines instead of 30, no more than that.
You add comments because your code is becoming unreadable because you made it take so much place.
You are creating variable when you don't need to, there is no reason to be that verbose.

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+            .responseTimeout(Duration.ofSeconds(6))
+            .build()
+            .post()
+            .uri(endpoint)
+            .bodyValue(codeInput)
+            .exchange()
+            .expectStatus()
+            .isOk()
+            .expectBody(JShellResult.class)
+            .value((JShellResult evalResult) -> assertThat(evalResult).isNotNull())
```
Comment (Alathreon) [inline JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:78]: You are checking twice that it is not null.

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+            .returnResult()
+            .getResponseBody();
+
+        assertThat(result).isNotNull();
```
Comment (Alathreon) [inline JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:82]: You are checking twice that it is not null, only keep this one.

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+
+        assertThat(testEval(testEvalId, secondCodeExpression)).isEqualTo(secondCodeExpectedResult);
+    }
+
+    private JShellResult testEval(String testEvalId, String codeInput) {
+        final String endpoint =
+                String.join("/", ApiEndpoints.BASE, ApiEndpoints.EVALUATE, testEvalId);
+
+        JShellResult result = this.webTestClient.mutate()
+            .responseTimeout(Duration.ofSeconds(6))
```
Comment (Alathreon) [inline JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:69]: 6s is lower than the api limit, so either increase this, or find a way to use a different config specifically for the tests. I would prefer this later approach.

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+            .value((JShellResult evalResult) -> assertThat(evalResult).isNotNull())
+            .returnResult()
+            .getResponseBody();
+
+        assertThat(result).isNotNull();
+
+        return result;
+    }
+
+    private static JShellResult getJShellResultDefaultInstance(JShellSnippetResult snippetResult) {
```
Comment (Alathreon) [inline JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:87]: This method is useless, the name is nearly as long as its code, its name is very confusing and doesn't help at all.
And you would need custom values most of the time anyway.

File: JShellWrapper/src/test/java/JShellWrapperTest.java [hunk 23]
```diff
@@ -49,12 +49,10 @@ void testHelloWorld() {
...
-        evalTest(
-                """
-            eval
-            1
-            "Hello world!\"""",
-                """
+        evalTest("""
+                eval
+                1
+                "Hello world!\"""", """
```
Comment (Alathreon) [inline JShellWrapper/src/test/java/JShellWrapperTest.java:55]: Did you check that the tests still work ?

File: JShellWrapper/src/test/java/JShellWrapperTest.java [hunk 24]
```diff
@@ -67,12 +65,10 @@ void testExpressionResult() {
...
-        evalTest(
-                """
-            eval
-            1
-            2+2""",
-                """
+        evalTest("""
+                eval
+                1
+                2+2""", """
```
Comment (Alathreon) [inline JShellWrapper/src/test/java/JShellWrapperTest.java:71]: Did you check that the tests still work ?

**Review by Alathreon — CHANGES_REQUESTED** [review]

File: JShellAPI/src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 14 above)
Comment (surajkumar) [reply JShellAPI/src/main/resources/META-INF/additional-spring-configuration-metadata.json:8]: It's used by your IDE so you get proper linting in the context of Spring @Alathreon

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java (hunk 12 above)
Comment (firasrg) [reply JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java:34]: The `jshell-wrapper.image-name`  property is not under `jshellapi` in properties file

File: JShellAPI/src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 14 above)
Comment (firasrg) [reply JShellAPI/src/main/resources/META-INF/additional-spring-configuration-metadata.json:8]: I guess no problem here ? 😎

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java (hunk 17 above)
Comment (firasrg) [reply JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:23]: > Why do you need to specify this ?

Somebody may not notice the additional configs in _build.gradle_ file about tests

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java (hunk 12 above)
Comment (Alathreon) [reply JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java:34]: See https://github.com/Together-Java/JShellPlaygroundBackend/pull/50#discussion_r1760314383

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java (hunk 17 above)
Comment (Alathreon) [reply JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:23]: How can you not notice it ? You need to check build.gradle to see how the tests work.

File: JShellWrapper/src/test/java/JShellWrapperTest.java (hunk 23 above)
Comment (firasrg) [reply JShellWrapper/src/test/java/JShellWrapperTest.java:55]: i didn't touch this file, it was reformatted after i tried the spotless:apply command

File: JShellWrapper/src/test/java/JShellWrapperTest.java (hunk 24 above)
Comment (firasrg) [reply JShellWrapper/src/test/java/JShellWrapperTest.java:71]: as i said on [the other comment](https://github.com/Together-Java/JShellPlaygroundBackend/pull/50#discussion_r1769172956)

File: JShellAPI/src/main/resources/application.yaml [hunk 25]
```diff
@@ -20,6 +20,9 @@ jshellapi:
   dockerResponseTimeout: 60
   dockerConnectionTimeout: 60
 
+  # JShellWrapper related
+  jshellwrapper-imageName: ${jshellWrapperImageName}
```
Comment (Alathreon) [inline JShellAPI/src/main/resources/application.yaml:24]: Please follow the conventions : camelCase.

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/dto/JShellResult.java [hunk 26]
```diff
@@ -9,4 +9,8 @@ public record JShellResult(List<JShellSnippetResult> snippetsResults,
     public JShellResult {
         snippetsResults = List.copyOf(snippetsResults);
     }
+
+    public JShellResult(JShellSnippetResult snippetsResultList) {
+        this(List.of(snippetsResultList), null, false, "");
+    }
```
Comment (Alathreon) [inline JShellAPI/src/main/java/org/togetherjava/jshellapi/dto/JShellResult.java:15]: This isn't used outside of a single test, so probably not a good idea to have.

File: JShellWrapper/src/test/java/JShellWrapperTest.java (hunk 23 above)
Comment (Alathreon) [reply JShellWrapper/src/test/java/JShellWrapperTest.java:55]: Still try the tests please

File: JShellWrapper/src/test/java/JShellWrapperTest.java (hunk 24 above)
Comment (Alathreon) [reply JShellWrapper/src/test/java/JShellWrapperTest.java:71]: Same

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java [hunk 27]
```diff
@@ -59,22 +62,30 @@ private void cleanupLeftovers(UUID currentId) {
 
     public String spawnContainer(long maxMemoryMegs, long cpus, @Nullable String cpuSetCpus,
             String name, Duration evalTimeout, long sysoutLimit) throws InterruptedException {
-        String imageName = "togetherjava.org:5001/togetherjava/jshellwrapper";
+
+        String[] imageNameParts = this.jshellWrapperImageName.split(":master");
+
+        if (imageNameParts.length != 1) {
+            throw new IllegalArgumentException("invalid jshellWrapper image name");
+        }
```
Comment (Alathreon) [inline JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java:70]: This should probably be checked before, in the config constructor

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/rest/ApiEndpoints.java [hunk 28]
```diff
@@ -0,0 +1,16 @@
+package org.togetherjava.jshellapi.rest;
+
+/**
+ * This class holds endpoints mentioned in controllers. The main objective is to keep endpoints
+ * synchronized with testing classes.
```
Comment (Alathreon) [inline JShellAPI/src/main/java/org/togetherjava/jshellapi/rest/ApiEndpoints.java:5]: Don't use
> This class

And don't mention the tests, a class outside of tests shouldn't exist for the purpose of tests.

File: JShellAPI/README.MD [hunk 29]
```diff
@@ -87,18 +87,28 @@ The following startup scripts id can be used :
...
+#### jshellapi.evalTimeoutSeconds
 The timeout of an evaluation, in seconds, see [Eval timeout](#Eval-timeout)
-### jshellapi.maxAliveSessions
+#### jshellapi.maxAliveSessions
 The maximum number of alive sessions, see [Errors](#Errors)
-### jshellapi.dockerMaxRamMegaBytes
+#### jshellapi.dockerMaxRamMegaBytes
 The maximum ram allocated per container, in megabytes.
-### jshellapi.dockerCPUsUsage
+#### jshellapi.dockerCPUsUsage
```
Comment (Alathreon) [inline JShellAPI/README.MD:103]: Create another PR for this, this is outside the scope of this PR.

File: JShellAPI/README.MD [hunk 30]
```diff
@@ -87,18 +87,28 @@ The following startup scripts id can be used :
...
+#### jshellapi.schedulerSessionKillScanRate
+The rate at which the session killer will check and delete session, in seconds, see [Session timeout](#Session-timeout).
+
+### Testing
+**JShellWrapper** Docker image lifecycle is handled during tests.
+
+The image name is injected from the root [build.gradle](../build.gradle) file, to this project's [build.gradle](build.gradle) file and also to [application.yaml](src/main/resources/application.yaml).
+
+- JShellWrapper Docker Image is built before the tests run.
+- **Container & Cleanup**: After the tests finish, the container and image are removed to ensure a clean environment.
```
Comment (Alathreon) [inline JShellAPI/README.MD:114]: Like I said last time, this is the wrong place for this, this is the readme, not the doc.
But, you didn't mention how to run the tests, which is important to have in the readme. And if you add it, prefer adding it to the readme in the parent project, JShellPlaygroundBackend + how to run JShellWrapper if there is no mention of it.

**Review by Alathreon — CHANGES_REQUESTED** [review]

File: JShellAPI/src/main/resources/application.yaml (hunk 25 above)
Comment (firasrg) [reply JShellAPI/src/main/resources/application.yaml:24]: I wanted to keep `jshellwrapper` like the top level property `jshellapi` which is defined lowercase

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java (hunk 27 above)
Comment (firasrg) [reply JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java:70]: It is in the constructor !

Comment (firasrg) [reply JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java:70]: I don't see a reason to do that the parts aren't needed in the constructor!

File: JShellWrapper/src/test/java/JShellWrapperTest.java (hunk 23 above)
Comment (firasrg) [reply JShellWrapper/src/test/java/JShellWrapperTest.java:55]: no issues in wrapper tests ✅

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java (hunk 27 above)
Comment (Alathreon) [reply JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java:70]: But it isn't the responsibility of this class to check this

File: JShellAPI/src/main/resources/application.yaml (hunk 25 above)
Comment (Alathreon) [reply JShellAPI/src/main/resources/application.yaml:24]: Oh yeah, well, jshellapi should have been called jshell-api, so we will fix this later.

File: JShellAPI/README.MD [hunk 31]
```diff
@@ -101,4 +101,4 @@ The maximum ram allocated per container, in megabytes.
 ### jshellapi.dockerCPUsUsage
 The cpu configuration of each container, see [--cpus option of docker](https://docs.docker.com/config/containers/resource_constraints/#cpu).
 ### jshellapi.schedulerSessionKillScanRate
-The rate at which the session killer will check and delete session, in seconds, see [Session timeout](#Session-timeout).
\ No newline at end of file
+The rate at which the session killer will check and delete session, in seconds, see [Session timeout](#Session-timeout).
```
Comment (Alathreon) [inline JShellAPI/README.MD:104]: What ?

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/Utils.java [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+        if (!StringUtils.hasText(imageName)
+                || !imageName.endsWith(Config.JSHELL_WRAPPER_IMAGE_NAME_TAG)) {
+            return false;
+        }
+
+        final String imageNameFirstPart = imageName.split(Config.JSHELL_WRAPPER_IMAGE_NAME_TAG)[0];
+
+        return StringUtils.hasText(imageNameFirstPart);
+    }
+}
```
Comment (Alathreon) [inline JShellAPI/src/main/java/org/togetherjava/jshellapi/Utils.java:18]: Move this to a static private method in Config, since it's only used once and should only be handled by Config.

File: JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java [hunk 33]
```diff
@@ -59,22 +62,27 @@ private void cleanupLeftovers(UUID currentId) {
 
     public String spawnContainer(long maxMemoryMegs, long cpus, @Nullable String cpuSetCpus,
             String name, Duration evalTimeout, long sysoutLimit) throws InterruptedException {
-        String imageName = "togetherjava.org:5001/togetherjava/jshellwrapper";
+
+        String[] imageNameParts =
+                this.jshellWrapperImageName.split(Config.JSHELL_WRAPPER_IMAGE_NAME_TAG);
+
+        String baseImageName = imageNameParts[0];
```
Comment (Alathreon) [inline JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java:69]: Do this in the constructor.

File: JShellAPI/src/main/resources/application.yaml (hunk 25 above)
Comment (Alathreon) [reply JShellAPI/src/main/resources/application.yaml:24]: Still not fixed

File: README.md [hunk 34]
```diff
@@ -47,3 +47,11 @@ JShellAPI is a REST API, and whenever some code is received, it will create a se
...
 ## How to use JShellApi ?
 See [JShellAPI README](JShellAPI/README.MD)
+
+## Testing
+**JShellWrapper** Docker image lifecycle is handled during tests.
+
+The image name is injected from the root [build.gradle](./build.gradle) file, to JShellAPI project's [build.gradle](./JShellAPI/build.gradle) file and also to its [application.yaml](./JShellAPI/src/main/resources/application.yaml).
+
+- JShellWrapper Docker Image is built before the tests run.
+- **Container & Cleanup**: After the tests finish, the container and image are removed to ensure a clean environment.
```
Comment (Alathreon) [inline README.md:57]: How many times I have to repeat this, readme isn't documentation nor user manual, you are supposed to tell the user how to run the tests, not how the project works.

**Review by Alathreon — CHANGES_REQUESTED** [review]

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java [hunk 35]
```diff
@@ -0,0 +1,74 @@
...
+ */
+@ContextConfiguration(classes = Main.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class JShellApiTests {
+
+    @Autowired
+    private WebTestClient webTestClient;
+
+    @Autowired
+    private Config appConfig;
```
Comment (Alathreon) [inline JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:32]: How do you handle custom config ? Like if for one test, you need to use custom config ?

**Review by Alathreon — CHANGES_REQUESTED** [review]

File: JShellAPI/README.MD (hunk 31 above)
Comment (firasrg) [reply JShellAPI/README.MD:104]: huh i dunno what's this, i'll revert

File: README.md (hunk 34 above)
Comment (firasrg) [reply README.md:57]: oh *** i forgot it

File: JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java (hunk 35 above)
Comment (firasrg) [reply JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:32]: I thought about that, but I don't think that would be a good idea as it's an integration-test, and it must verify what exactly happens in real scenario (with the real config).

Comment (Alathreon) [reply JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:32]: The problem is that the real world config has timers from the minute to the dozen of minutes, you do not want a test to last for 30 minute because you wanted to check if a session correctly disapear after 30 minutes...

Comment (firasrg) [reply JShellAPI/src/test/java/org/togetherjava/jshellapi/JShellApiTests.java:32]: in that case we may create a common config and separate config

File: JShellAPI/README.MD (hunk 31 above)
Comment (firasrg) [inline JShellAPI/README.MD:104]: FYI, this change is due to `spotlessApply` task

**Review by Alathreon — APPROVED** [review]
